skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,565 @@
+                                        x_col_start_j: x_col_end_j, :],
+                                        w, s)
+                    weight_sum += weight
+                    for color in range(n_ch):
+                        new_values[color] += weight * padded[x_row + i,
+                                                             x_col + j, color]
+            for color in range(n_ch):
+                result[x_row, x_col, color] = new_values[color] / weight_sum
+    return result[offset:-offset, offset:-offset]
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def _nl_means_denoising_3d(image, int s=7,
+            int d=13, float h=0.1):
+    """
+    Perform non-local means denoising on 3-D array
+
+    Parameters
+    ----------
+    image: ndarray
+        input data to be denoised
+    s: int, optional
+        size of patches used for denoising
+    d: int, optional
+        maximal distance in pixels where to search patches used for denoising
+    h: float, optional
+        cut-off distance (in gray levels)
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int n_pln, n_row, n_col
+    n_pln, n_row, n_col = image.shape
+    cdef int offset = s / 2
+    # padd the image so that boundaries are denoised as well
+    cdef DTYPE_t [:, :, ::1] padded = np.ascontiguousarray(util.pad(
+                                        image.astype(np.float32),
+                                        offset, mode='reflect'))
+    cdef DTYPE_t [:, :, ::1] result = padded.copy()
+    cdef float A = ((s - 1.) / 4.)
+    cdef float new_value
+    cdef float weight_sum, weight
+    xg_pln, xg_row, xg_col = np.mgrid[-offset: offset + 1,
+                                      -offset: offset + 1,
+                                      -offset: offset + 1]
+    cdef DTYPE_t [:, :, ::1] w = np.ascontiguousarray(np.exp(
+                            - (xg_pln ** 2 + xg_row ** 2 + xg_col ** 2) /
+                            (2 * A ** 2)).astype(np.float32))
+    cdef float distance
+    cdef int x_pln, x_row, x_col, i, j, k
+    cdef int x_pln_start, x_pln_end, x_row_start, x_row_end, \
+             x_col_start, x_col_end
+    cdef int x_pln_start_i, x_pln_end_i, x_row_start_j, x_row_end_j, \
+             x_col_start_k, x_col_end_k
+    w = 1. / (np.sum(w) * h ** 2) * w
+    # Coordinates of central pixel and patch bounds
+    for x_pln in range(offset, n_pln + offset):
+        x_pln_start = x_pln - offset
+        x_pln_end = x_pln + offset + 1
+        for x_row in range(offset, n_row + offset):
+            x_row_start = x_row - offset
+            x_row_end = x_row + offset + 1
+            for x_col in range(offset, n_col + offset):
+                x_col_start = x_col - offset
+                x_col_end = x_col + offset + 1
+                new_value = 0
+                weight_sum = 0
+                # Coordinates of test pixel and patch bounds
```

Comment:
It [looks like](http://docs.scipy.org/doc/numpy/reference/arrays.nditer.html#putting-the-inner-loop-in-cython) `np.nditer` can be quite performant in Cython! (when provided with dtype info, which we can do.)